File: apps/viewer/src/types/fsm.ts
import type { DefinedAt, Expression } from "./source";

export type Fsm = {
  id: string;
  name: string;
  context?: string;
  states?: State[];
  events?: EventDef[];
  guards?: Guard[];
  actions?: Action[];
  transitions?: Transition[];
  defined_at?: DefinedAt;
  visibility?: string;
  tags?: string[];
};

export type State = {
  id: string;
  kind: string;
  initial?: boolean;
  terminal?: boolean;
  terminal_semantics?: string;
  defined_at?: DefinedAt;
  visibility?: string;
  tags?: string[];
};

export type EventDef = {
  id: string;
  name?: string;
  kind?: string;
};

export type Guard = {
  id: string;
  name?: string;
  kind?: string;
  capability?: string;
  expression?: Expression;
  input?: string;
  defined_at?: DefinedAt;
  visibility?: string;
  tags?: string[];
};

export type Action = {
  id: string;
  name?: string;
  kind?: string;
  capability?: string;
  command?: string;
  emits?: string[];
  expression?: Expression;
  depends_on?: string[];
  defined_at?: DefinedAt;
  visibility?: string;
  tags?: string[];
};

export type Transition = {
  id: string;
  from: string;
  to: string;
  on?: string;
  guards?: string[];
  actions?: string[];
  requires?: string[];
  defined_at?: DefinedAt;
  visibility?: string;
  tags?: string[];
};

File: crates/dslraid-core/src/model/fsm.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::{Action, DefinedAt, Event, Guard, State, Transition};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Fsm {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub context: Option<String>,
    #[serde(default)]
    pub states: Vec<State>,
    #[serde(default)]
    pub events: Vec<Event>,
    #[serde(default)]
    pub guards: Vec<Guard>,
    #[serde(default)]
    pub actions: Vec<Action>,
    #[serde(default)]
    pub transitions: Vec<Transition>,
    #[serde(default)]
    pub defined_at: Option<DefinedAt>,
    #[serde(default)]
    pub visibility: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub metadata: Option<Value>,
}
